v4/supervisor.go

```diff
@@ -0,0 +1,956 @@
+			})
+		}
+		s.m.Lock()
+		s.unstoppedServiceReport = report
+		s.m.Unlock()
+		return report
+	}
+}
+
+// String implements the fmt.Stringer interface.
+func (s *Supervisor) String() string {
+	return s.Name
+}
+
+// sendControl abstracts checking for the supervisor to still be running
+// when we send a message. This prevents blocking when sending to a
+// cancelled supervisor.
+func (s *Supervisor) sendControl(sm supervisorMessage) error {
+	var doneChan <-chan struct{}
+	s.ctxMutex.Lock()
+	if s.ctx == nil {
+		s.ctxMutex.Unlock()
+		return ErrSupervisorNotStarted
+	}
+	doneChan = s.ctx.Done()
+	s.ctxMutex.Unlock()
+
+	select {
+	case s.control <- sm:
+		return nil
+	case <-doneChan:
+		return ErrTimeout
+	}
+}
+
+/*
+Remove will remove the given service from the Supervisor, and attempt to Stop() it.
+The ServiceID token comes from the Add() call. This returns without waiting
+for the service to stop.
+*/
+func (s *Supervisor) Remove(id ServiceToken) error {
+	sID := supervisorID(id.id >> 32)
+	if sID != s.id {
+		return ErrWrongSupervisor
+	}
+	err := s.sendControl(removeService{serviceID(id.id & 0xffffffff), nil})
+	if err == ErrTimeout {
+		// No meaningful error handling if the supervisor is stopped.
+		return nil
+	}
+	return err
+}
+
+/*
+RemoveAndWait will remove the given service from the Supervisor and attempt
+to Stop() it. It will wait up to the given timeout value for the service to
+terminate. A timeout value of 0 means to wait forever.
+
+If a nil error is returned from this function, then the service was
+terminated normally. If either the supervisor terminates or the timeout
+passes, ErrTimeout is returned. (If this isn't even the right supervisor
+ErrWrongSupervisor is returned.)
+*/
+func (s *Supervisor) RemoveAndWait(id ServiceToken, timeout time.Duration) error {
+	sID := supervisorID(id.id >> 32)
+	if sID != s.id {
+		return ErrWrongSupervisor
+	}
+
+	var timeoutC <-chan time.Time
+
+	if timeout > 0 {
+		timer := time.NewTimer(timeout)
+		defer timer.Stop()
+		timeoutC = timer.C
+	}
+
+	notificationC := make(chan struct{})
+
+	sentControlErr := s.sendControl(removeService{serviceID(id.id & 0xffffffff), notificationC})
+
+	if sentControlErr != nil {
+		return sentControlErr
+	}
+
+	select {
+	case <-notificationC:
+		// normal case; the service is terminated.
+		return nil
+
+	// This occurs if the entire supervisor ends without the service
+	// having terminated, and includes the timeout the supervisor
+	// itself waited before closing the liveness channel.
+	case <-s.ctx.Done():
+		return ErrTimeout
+
+	// The local timeout.
+	case <-timeoutC:
+		return ErrTimeout
+	}
+}
+
+/*
+
+Services returns a []Service containing a snapshot of the services this
+Supervisor is managing.
+
+*/
+func (s *Supervisor) Services() []Service {
+	ls := listServices{make(chan []Service)}
+
+	if s.sendControl(ls) == nil {
+		return <-ls.c
+	}
+	return nil
+}
+
+var currentSupervisorIDL sync.Mutex
```

Comment:
Replace with atomic.AddUint32 ?

Reply:
To be honest, I've never gotten a clear sense of exactly what contract the atomic functions provide, in particular, the read guarantees they provide, so I'm a bit nervous about putting this out in such a basic library.

(In particular, I'd really like to know for sure for non-suture-related code that an atomic pointer swap's result can be safely read without any atomic operations, but I can't find any documentation or person willing to guarantee that. I've been doing it anyhow in some code and it seems to work, and it seems like there can't be any other purpose for the operation since there is no AtomicRead, but I'd really like it to be stated somewhere.)

Reply:
Hmm, it would be absolutely astonishing if the return value of `CompareAndSwap`, `AddUint32`, etc required synchronization.  It would practically defeat the purpose of these functions.

> there is no AtomicRead

There's the `atomic.Load*` [family](https://golang.org/pkg/sync/atomic/#LoadInt32), but they seem to be the counterparty to `atomic.Store*`.

Personally, I feel comfortable taking that bet (in fact, I still can't bring myself to consider it a bet at all), but I also can't imagine the mutex presents a meaningful bottleneck, either.  I guess there's no harm in playing it safe.